README.md
# MCScanner
Now re-made in Java, and not Python!

## Requirements
* [`JDK 17`][adopt17]
* MongoDB database (required for storing IPs)
* Stable internet connection
* A good computer (with a good CPU)

## Run
### macOS + Linux + Windows
1. Click [here][latest] to download the latest tested, safe, and stable build.
2. Run the `.jar` file with `java -jar /path/to/MCScanner.jar` from Command Prompt or Terminal.

> **Note**<br>
> If you're on macOS, you can just double-click on the `.jar` file.

## Build
### macOS + Linux + Windows
1. Click [here][latest-source] to download untested, unsafe & unstable source code.
2. Open it in IntelliJ IDEA.
3. Build the project with `Build > Build Project`.
4. Run the project with `Run > Run 'MCScanner [run]'`.

## Contributing
### General
If you want to contribute to this project, you can fork this repository, make your changes, and then make a pull request!
### Languages
To contribute to the languages, fork the repository, analyse the `en-GB.json` file, make a new file with your language locale (see [this][lang]), translate the values (not the keys), and then make a pull request!
Please don't use Google Translate :(

## Changelogs
All notable changes to this project will be documented in [`CHANGELOG.md`][changes].

## Warning
> **Warning**<br>
> This program was made for **educational purposes only**, and is **not meant to be used in malicious ways**. 
> **Do not DDOS the IPs** or do *anything* malicious with the IPs.
> The only thing this program is supposed to be used for is finding random Minecraft servers and joining them, to have fun!

Oh, and **don't use this program for griefing**. Please.

[adopt17]: https://adoptium.net/en-GB/download/

[changes]: https://github.com/StupidRepo/MCScanner/blob/main/CHANGELOG.md
[latest]: https://github.com/StupidRepo/MCScanner/releases/latest/download/MCScanner.jar
[latest-source]: https://github.com/StupidRepo/MCScanner/archive/refs/heads/main.zip

[lang]: https://www.oracle.com/java/technologies/javase/jdk17-suported-locales.html

CHANGELOG.md
# Changes
All notable changes to this project will be documented in `CHANGELOG.md`.

## Added
* A [LanguageHandler][langhand] for handling strings in different languages. If you want to add a new language, read [this][readme-lang].

## Modified
* Fixed `dropdown.SERVERLIST.*` strings saying 'Sort By' instead of 'Search By'.

## Removed
Nothing has been removed.

## TODOs
- [ ] Optimise IP generation and initial scanning code.
- [ ] Optimise code in general.
- [ ] Separate tool for viewing servers in DB. (code is already here (`ServerList` class), just need to make it separate)
- [x] ~~Make it save what IP it got to when quitting so that it can resume from that IP on next startup.~~
- [x] ~~Add a GUI for viewing servers in a nice friendly grid.~~

[readme-lang]: https://github.com/StupidRepo/MCScanner/tree/main#languages
[langhand]: https://github.com/StupidRepo/MCScanner/blob/main/src/com/stupidrepo/mcscanner/language/LanguageHandler.java

src/com/stupidrepo/mcscanner/MCScanner.java
package com.stupidrepo.mcscanner;

import com.stupidrepo.mcscanner.language.LanguageHandler;
import org.bson.Document;

import javax.swing.*;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableModel;
import javax.swing.table.TableRowSorter;
import java.awt.*;
import java.awt.event.ActionListener;
import java.awt.event.WindowEvent;
import java.io.*;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.PatternSyntaxException;

public class MCScanner {
    private static int offsetI = 1;
    private static int offsetJ = 0;
    private static int offsetK = 0;
    private static int offsetL = 0;
    private static boolean stopping = false;

    public static LanguageHandler lang;

    public static void main(String[] args) {
        AtomicInteger threads = new AtomicInteger(1024);
        int timeout = 1000;
        int maxRange = 255;
        int port = 25565;

        Logger logger = Logger.getLogger("com.stupidrepo.mcscanner");

        float version = 1.21f;

        AtomicReference<String> uri = new AtomicReference<>("mongodb://localhost:27017");

        lang = new LanguageHandler(Locale.forLanguageTag("en-gb"));

        PopupHandler threadsPopup = new PopupHandler(lang.get("question.THREADS"), "1024", "OK");
        threadsPopup.showAndWait();

        try {
            threads.set(Integer.parseInt(threadsPopup.responseText));
        } catch (NumberFormatException e) {
            logger.log(Level.SEVERE, "Invalid thread count.");
        }

        PopupHandler mongoDBURIPopup = new PopupHandler(lang.get("question.MONGO"), "mongodb://localhost:27017", "Done");
        mongoDBURIPopup.showAndWait();

        uri.set(mongoDBURIPopup.responseText);

        DatabaseHandler databaseHandler = new DatabaseHandler(uri.get());

        logger.log(Level.INFO, "Scanning IPs...");

        JFrame frame = new JFrame(lang.get("text.TITLE").formatted(version));
        frame.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        frame.setSize(300, 100);
        frame.setLayout(new BorderLayout());

        ArrayList < Thread > threadList = new ArrayList < Thread > ();

        JLabel scannedLabel = new JLabel(lang.get("text.SCANNED").formatted(0));
        scannedLabel.setHorizontalAlignment(0);

        frame.add(scannedLabel, "Center");

        long scanned = 0;

        frame.addWindowListener(new java.awt.event.WindowAdapter() {
            @Override
            public void windowClosing(java.awt.event.WindowEvent windowEvent) {
                stopping = true;
                logger.log(Level.INFO, "Stopping threads...");
                for (Thread nextThread: threadList) {
                    try {
                        nextThread.join();
                    } catch (InterruptedException timeouter) {
                        // Timeout or something idk
                    }
                }
                logger.log(Level.INFO, "Making an '.mcscanner' file...");
                try {
                    BufferedWriter writer = new BufferedWriter(new FileWriter(".mcscanner"));
                    writer.write(offsetI + "\n" + offsetJ + "\n" + offsetK + "\n" + offsetL);
                    writer.close();
                } catch (IOException e) {
                    logger.log(Level.SEVERE, "Failed to write '.mcscanner'!");
                }
                logger.log(Level.INFO, "Exiting...");
                System.exit(0);
            }
        });

        ServerList serverList = new ServerList(databaseHandler, lang);

        JButton viewServersButton = new JButton(lang.get("button.SERVERLIST"));

        viewServersButton.addActionListener(e -> {
            serverList.toggleGUI();
        });

        frame.add(viewServersButton, "North");

        frame.setVisible(true);

        File offsetFile = new File(".mcscanner");
        if (offsetFile.exists()) {
            try {
                logger.log(Level.INFO, "Found '.mcscanner'.");
                BufferedReader reader = new BufferedReader(new FileReader(offsetFile));
                offsetI = Integer.parseInt(reader.readLine());
                offsetJ = Integer.parseInt(reader.readLine());
                offsetK = Integer.parseInt(reader.readLine());
                offsetL = Integer.parseInt(reader.readLine());
                reader.close();
            } catch (IOException e) {
                logger.log(Level.SEVERE, "Failed to read '.mcscanner'.");
            }
            logger.log(Level.INFO, "Continuing from " + offsetI + "." + offsetJ + "." + offsetK + "." + offsetL + "...");
        }

        int thisOffsetI = offsetI;
        int thisOffsetJ = offsetJ;
        int thisOffsetK = offsetK;
        int thisOffsetL = offsetL;
        for (int i = thisOffsetI; i <= maxRange; ++i) {
            if(stopping) {
                break;
            } else {
                offsetI = i;
            }
            for (int j = thisOffsetJ; j <= 255; ++j) {
                if(stopping) {
                    break;
                } else {
                    offsetJ = j;
                }
                for (int k = thisOffsetK; k <= 255; ++k) {
                    if(stopping) {
                        break;
                    } else {
                        offsetK = k;
                    }
                    for (int l = thisOffsetL; l <= 255; ++l) {
                        String ip = "...";
                        if(stopping) {
                            break;
                        } else {
                            offsetL = l;
                            ip = i + "." + j + "." + k + "." + l;

                            ScannerThread scannerThread = new ScannerThread(ip, port, timeout, databaseHandler);
                            Thread scanThread = new Thread(scannerThread);
                            threadList.add(scanThread);
                            scanThread.start();
                        }

                        if (threadList.size() >= threads.get()) {
                            for (Thread nextThread: threadList) {
                                try {
                                    nextThread.join();
                                    ++scanned;
                                    scannedLabel.setText(lang.get("text.SCANNED").formatted(scanned));
                                } catch (InterruptedException timeout2) {
                                    // Timed out or smth
                                }
                            }
                            threadList.clear();
                        }
                    }
                    thisOffsetL = 0;
                }
                thisOffsetK = 0;
            }
            thisOffsetJ = 0;
        }

        for (Thread nextThreadAgain: threadList) {
            try {
                nextThreadAgain.join();
                ++scanned;
                scannedLabel.setText(lang.get("text.SCANNED").formatted(scanned));
            } catch (InterruptedException timeout1) {
                // Timeout, again!
            }
        }

        if(!stopping) {
            frame.setVisible(false);
            frame.dispatchEvent(new WindowEvent(frame, 201));
            logger.log(Level.INFO, "Scan completed!");
        }
    }
}

class ScannerThread implements Runnable {
    private final String ip;
    private final int port;
    private final int timeout;
    private final DatabaseHandler dbHandler;

    public boolean didHit = false;

    public ScannerThread(String ip, int port, int timeout, DatabaseHandler dbHandler) {
        this.ip = ip;
        this.port = port;
        this.timeout = timeout;
        this.dbHandler = dbHandler;
    }

    public void run() {
        try {
            Socket socket = new Socket();
            socket.connect(new InetSocketAddress(this.ip, this.port), this.timeout);

            InputStream inputStream = socket.getInputStream();
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream, StandardCharsets.UTF_16BE);

            OutputStream outputStream = socket.getOutputStream();

            outputStream.write(new byte[] {
                    (byte) 0xFE, (byte) 0x01
            });

            socket.setSoTimeout(this.timeout);

            int packetId = inputStream.read();

            if (packetId != 0xFF) {
                socket.close();
                throw new IOException("Invalid packet: (" + packetId + ")");
            }

            int length = inputStreamReader.read();

            if (length == -1 || length == 0) {
                socket.close();
                throw new IOException("Premature end of stream/invalid String length.");
            }

            char[] chars = new char[length];

            if (inputStreamReader.read(chars, 0, length) != length) {
                socket.close();
                throw new IOException("Premature end of stream.");
            }

            String string = new String(chars);

            String[] data;
            if (string.startsWith("§")) {
                data = string.split("\0");

                if(dbHandler.isIPInDB(ip)) {
                    dbHandler.updateServerByIPInDB(ip, data[2], data[3], Integer.parseInt(data[5]));
                } else {
                    dbHandler.writeDetailsToDB(ip, data[2], data[3], Integer.parseInt(data[5]));
                }
            } else {
                data = string.split("§");

                if(dbHandler.isIPInDB(ip)) {
                    dbHandler.updateServerByIPInDB(ip, "1.6<=", data[0], Integer.parseInt(data[2]));
                } else {
                    dbHandler.writeDetailsToDB(ip, "1.6<=", data[0], Integer.parseInt(data[2]));
                }
            }

            socket.close();

            didHit = true;
        } catch (IOException var8) {
            // No response/invalid response/timeout/port accidentally left open
        }
    }
}

class ServerList {
    private final DatabaseHandler dbHandler;
    private final LanguageHandler lang;
    private final JFrame frame;

    public ServerList(DatabaseHandler dbHandler, LanguageHandler lang) {
        this.lang = lang;
        this.dbHandler = dbHandler;
        this.frame = new JFrame(lang.get("text.SERVERLIST.TITLE").formatted(0));
        this.frame.setDefaultCloseOperation(WindowConstants.HIDE_ON_CLOSE);
        this.frame.setSize(720, 500);
        this.frame.setLayout(new BorderLayout());
        
        JTable table = new JTable();
        table.setModel(new DefaultTableModel(
            new Object[][] {
            },
            new String[] {
                lang.get(this.lang.get("text.SERVERLIST.IP")), this.lang.get("text.SERVERLIST.MOTD"), this.lang.get("text.SERVERLIST.VERSION"), this.lang.get("text.SERVERLIST.MAX_PLAYERS")
            }
        ));
        table.getColumnModel().getColumn(0).setPreferredWidth(100);
        table.getColumnModel().getColumn(1).setPreferredWidth(200);
        table.getColumnModel().getColumn(2).setPreferredWidth(50);
        table.getColumnModel().getColumn(3).setPreferredWidth(50);
        table.setAutoResizeMode(JTable.AUTO_RESIZE_LAST_COLUMN);
        table.setFillsViewportHeight(true);
        table.setRowHeight(20);
        table.setRowSelectionAllowed(false);
        table.setDefaultEditor(Object.class, null);
        table.setColumnSelectionAllowed(false);
        table.setCellSelectionEnabled(false);
        table.getTableHeader().setReorderingAllowed(false);
        table.getTableHeader().setResizingAllowed(false);
        
        JScrollPane scrollPane = new JScrollPane(table);
        scrollPane.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
        scrollPane.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);

        JTextField searchBar = new JTextField();
        searchBar.setHorizontalAlignment(0);
        searchBar.setToolTipText("Search");
        searchBar.addFocusListener(new PlaceholderText("Search", searchBar).getFocusAdapter());

        JComboBox<String> searchBy = new JComboBox<String>();
        searchBy.addItem(lang.get("dropdown.SERVERLIST.IP"));
        searchBy.addItem(lang.get("dropdown.SERVERLIST.MOTD"));
        searchBy.addItem(lang.get("dropdown.SERVERLIST.VERSION"));
        searchBy.addItem(lang.get("dropdown.SERVERLIST.MAX_PLAYERS"));
        searchBy.setSelectedIndex(2);

        searchBar.addActionListener(e -> {
            String text = searchBar.getText();
            if(text.length() > 0) {
                try {
                    ((TableRowSorter<TableModel>) table.getRowSorter()).setRowFilter(RowFilter.regexFilter(text, searchBy.getSelectedIndex()));
                } catch (PatternSyntaxException pse) {
                    JOptionPane.showMessageDialog(null, "Invalid search query/regex expression!", "Error", JOptionPane.ERROR_MESSAGE);
                }
            } else {
                ((TableRowSorter<TableModel>) table.getRowSorter()).setRowFilter(null);
            }
        });

        Timer timer = new Timer(10000, e -> {
            ((DefaultTableModel) table.getModel()).setRowCount(0);
            ArrayList < Document > documents1 = this.dbHandler.getServers();
            this.frame.setTitle(lang.get("text.SERVERLIST.TITLE").formatted(documents1.size()));
            for (Document document: documents1) {
                String ip = document.getString("ip");
                String motd = document.getString("motd");
                String version = document.getString("version");
                int players = document.getInteger("maxPlayers");

                ((DefaultTableModel) table.getModel()).addRow(new Object[] {
                    ip, motd, version, players
                });
            }
        });
        timer.setRepeats(true);
        timer.start();
        timer.getListeners(ActionListener.class)[0].actionPerformed(null);

        TableRowSorter < TableModel > sorter = new TableRowSorter<> (table.getModel());
        table.setRowSorter(sorter);

        this.frame.add(scrollPane, BorderLayout.CENTER);
        this.frame.add(searchBy, BorderLayout.SOUTH);
        this.frame.add(searchBar, BorderLayout.NORTH);
    }

    public boolean toggleGUI() {
        this.frame.setVisible(!this.frame.isVisible());
        return this.frame.isVisible();
    }
}

src/com/stupidrepo/mcscanner/PopupHandler.java
package com.stupidrepo.mcscanner;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JTextField;
import javax.swing.WindowConstants;
import java.awt.BorderLayout;
import java.awt.event.WindowEvent;

public class PopupHandler {
    private final String message;
    private final String defaultText;
    private final String buttonText;

    public String responseText;

    public PopupHandler(String message, String defaultText, String buttonText) {
        this.message = message;
        this.defaultText = defaultText;
        this.buttonText = buttonText;
    }

    public void showAndWait() {
        JFrame threadFrame = new JFrame("MCScanner");
        threadFrame.setDefaultCloseOperation(WindowConstants.DO_NOTHING_ON_CLOSE);
        threadFrame.setSize(500, 125);
        threadFrame.setLayout(new BorderLayout());

        JLabel threadLabel = new JLabel(this.message);
        threadLabel.setHorizontalAlignment(0);

        threadFrame.add(threadLabel, "North");

        JTextField threadField = new JTextField(this.defaultText);
        threadField.setHorizontalAlignment(0);

        threadFrame.add(threadField, "Center");

        JButton threadButton = new JButton(this.buttonText);
        threadFrame.add(threadButton, "East");

        threadButton.addActionListener(e -> {
            this.responseText = threadField.getText();
            threadFrame.setVisible(false);
            threadFrame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
            threadFrame.dispatchEvent(new WindowEvent(threadFrame, 201));
        });

        threadFrame.setVisible(true);

        while(threadFrame.isVisible()) {
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        threadFrame.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
    }
}


src/com/stupidrepo/mcscanner/DatabaseHandler.java
package com.stupidrepo.mcscanner;

import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;

import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;

public class DatabaseHandler {
    public MongoClient mainMongoClient;
    public MongoDatabase mainDatabase;
    private MongoCollection<Document> mainCollection;

    private final Logger logger = Logger.getLogger("com.stupidrepo.mcscanner");

    /**
     * Initiates a new DatabaseHandler.
     *
     * @param uri The URI to the MongoDB Database
     */
    public DatabaseHandler(String uri) {
        try {
            mainMongoClient = MongoClients.create(uri);
            mainDatabase = mainMongoClient.getDatabase("MCScanner");
            mainCollection = mainDatabase.getCollection("servers");
            logger.log(Level.INFO, "Connected to database: '" + mainDatabase.getName() + "'");
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to connect to database!");
        }
    }

    /**
     * Gets all the servers in the database.
     */
    public ArrayList<Document> getServers() {
        ArrayList<Document> servers = new ArrayList<>();
        try {
            for (Document doc : mainCollection.find()) {
                servers.add(doc);
            }
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to get servers from database!");
        }

        return servers;
    }

    /**
     * Gets the server count from the database.
     */
    public Long getServerCount() {
        try {
            return mainCollection.countDocuments();
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to get server count from database!");
            return null;
        }
    }

    /**
     * Writes an entry to the database.
     *
     * @param ip The IP of the server (<code>192.168.0.1</code>)
     * @param version The version of the server (<code>1.20.1</code>)
     * @param motd The MOTD of the server (<code>A Minecraft Server</code>)
     * @param maxPlayers The max players of the server (<code>20</code>)
     */
    public void writeDetailsToDB(String ip, String version, String motd, int maxPlayers) {
        try {
            mainCollection
                    .insertOne(
                            new Document("ip", ip)
                                    .append("version", version)
                                    .append("motd", motd)
                                    .append("maxPlayers", maxPlayers)
                    );
            logger.log(Level.INFO, "Added " + ip + " to database.");
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to add " + ip + " to database.");
        }
    }

    /**
     * Updates a server in the database.
     *
     * @param ip The IP of the server
     * @param version The version of the server
     * @param motd The MOTD of the server
     * @param maxPlayers The max players of the server
     */
    public void updateServerByIPInDB(String ip, String version, String motd, int maxPlayers) {
        try {
            mainCollection.updateOne(
                    new Document("ip", ip),
                    new Document("$set", new Document("version", version)
                            .append("motd", motd)
                            .append("maxPlayers", maxPlayers))
            );
            logger.log(Level.INFO, "Updated " + ip + " in database.");
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to update " + ip + " in database.");
        }
    }

    /**
     * Checks if a server is in the database.
     *
     * @param ip The IP of the server
     */
    public boolean isIPInDB(String ip) {
        try {
            Document myDoc = mainCollection.find(new Document("ip", ip)).first();
            return myDoc != null;
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Failed to check if " + ip + " is in database.");
            return false;
        }
    }
}

src/com/stupidrepo/mcscanner/PlaceholderText.java
package com.stupidrepo.mcscanner;

import javax.swing.*;
import java.awt.event.FocusAdapter;

public class PlaceholderText {
    private final String text;
    private String lastText = "";
    private final JTextField textField;

    public PlaceholderText(String text, JTextField textField) {
        this.text = text;
        this.textField = textField;
    }

    public FocusAdapter getFocusAdapter() {
        return new FocusAdapter() {
            public void focusGained(java.awt.event.FocusEvent evt) {
                textField.setText(lastText);
            }

            public void focusLost(java.awt.event.FocusEvent evt) {
                lastText = textField.getText();
                if (textField.getText().isEmpty()) {
                    textField.setText(text);
                }
            }
        };
    }
}


src/com/stupidrepo/mcscanner/language/LanguageHandler.java
package com.stupidrepo.mcscanner.language;

import org.bson.BsonString;
import org.bson.json.JsonObject;

import java.io.*;
import java.util.*;
import java.util.logging.Logger;

import java.util.logging.Level;

public class LanguageHandler {
    private final Hashtable<Object, Object> strings = new Hashtable<>();

    private Locale language;
    private final Locale fallbackLocale;

    private final Logger logger = Logger.getLogger("com.stupidrepo.mcscanner");

    public LanguageHandler(Locale fallback) {
        this.fallbackLocale = fallback;
        this.language = Locale.getDefault();
        this.loadStringsToDictionary();
    }

    public String get(String key) {
        if(this.strings.containsKey(key)) {
            return ((BsonString) this.strings.get(key)).getValue();
        } else {
            logger.log(Level.SEVERE, "No string found for key '%s'.".formatted(key));
            return key;
        }
    }

    public Locale getLanguage() {
        return this.language;
    }

    public Locale setLanguage(String language) {
        this.language = Locale.forLanguageTag(language);
        this.loadStringsToDictionary();
        return this.language;
    }

    public void loadStringsToDictionary() {
        ClassLoader classLoader = getClass().getClassLoader();
        InputStream is = classLoader.getResourceAsStream("lang/%s.json".formatted(this.language.toLanguageTag()));

        if (is == null) { is = classLoader.getResourceAsStream("lang/%s.json".formatted(this.fallbackLocale.toLanguageTag())); }
        if (is == null) { logger.log(Level.SEVERE, "No language file found for %s or %s!".formatted(this.language.toLanguageTag(), this.fallbackLocale.toLanguageTag())); } else {
            try {
                BufferedReader reader = new BufferedReader(new InputStreamReader(is));
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) { sb.append(line); }
                JsonObject json = new JsonObject(sb.toString());
                this.strings.putAll(json.toBsonDocument());
                logger.log(Level.INFO, "Loaded language file for %s!".formatted(this.language.toLanguageTag()));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}


main.py
import json
import masscan
from mcstatus import JavaServer
import random
import concurrent.futures
import pymongo
import time
import sys

servers_found = 0

threads = 4
seed = random.randrange(sys.maxsize)

text = False

rescan = False

print("WARNING! Using or modifying the code in any way to turn it into "
      "a DDos program is illegal!\nThis program was made for educational purposes, "
      "and should not be used in malicious ways! Do not use this program with malicious intent!")
print("I made this program for EDUCATIONAL PURPOSES only. If you try to use this program to DDos someone, "
      "then you are responsible for what happens. Not me!")
print("Please read the Minecraft terms and the EULA here:\n"
      "TERMS: https://www.minecraft.net/en-us/terms | EULA: https://www.minecraft.net/en-us/eula")
print("Using this program normally means: You get servers, connect to them via their IP, and play on them. "
      "(Don't use this program just to find servers to grief, you cheeky little-!)")
print("Remember, using this program maliciously is illegal! DO NOT DDos people!")
print("(Oh and also, if you are playing on a multiplayer server whilst running this, "
      "it has a chance to not detect servers and disconnect you from the server!)")
print()

while True:
    try:
        thinga = input("Do you want me to update already scanned servers (default: N, options: Y/N)? ")
        if thinga.lower() == 'y':
            rescan = True
        elif thinga.lower() == 'n':
            rescan = False
        break
    except:
        break

while True:
    try:
        threads = int(input("How many threads should I use (default & recommended: 4)? "))
        if threads > 32:
            print("Going above 32 threads will:\n- use so much of your router's bandwidth that it'll slow your "
                  "router down (and slow the program down)\n- slow down your computer "
                  "significantly\nOR\n- crash it.")
        break
    except:
        break

if not rescan:
    while True:
        try:
            seedLocal = input("What seed should the random generator use (default: system time)? ")
            if seedLocal.strip().__len__() != 0:
                seed = seedLocal
            break
        except:
            break

rng = random.Random(seed)

myclient = pymongo.MongoClient("mongodb://localhost:27017/")

mydb = myclient["MCScanner"]
mycol = mydb["servers"]

startCount = mycol.estimated_document_count()

A = list(range(1, 0xff))
B = list(range(1, 0xff))

rng.shuffle(A)
rng.shuffle(B)

if not rescan:
    ip_ranges = []
else:
    ip_ranges = [i['ip'] for i in mycol.find()]

if not rescan:
    for a in A:
        for b in B:
            ip_range = f"{a}.{b}.0.0/16"
            ip_ranges.append(ip_range)

def doPrintLoop():
    while True:
        count = mycol.estimated_document_count()
        if rescan:
            print(f'\rUpdating servers. Updated: {servers_found}/{startCount} '
                  f'{"server" if servers_found == 1 else "servers"}. '
                  f'({round(servers_found / count * 100, 2)}%)',
                  end=' ', flush=True)
        elif count == 0:
            print(f'\rScanning for servers. Found: {servers_found} '
                  f'{"server" if servers_found == 1 else "servers"}. '
                  f'Servers in DB at start: {startCount} '
                  f'{"server" if startCount == 1 else "servers"}, '
                  f'servers in DB now: {count} {"server" if count == 1 else "servers"}.',
                  end=' ', flush=True)
        else:
            print(f'\rScanning for servers. Found: {servers_found} '
                  f'{"server" if servers_found == 1 else "servers" }. '
                  f'Servers in DB at start: {startCount} '
                  f'{"server" if startCount == 1 else "servers" }, '
                  f'servers in DB now: {count} {"server" if count == 1 else "servers" }. '
                  f'(+{round((count - startCount) / count * 100, 2)}%)',
                  end=' ', flush=True)
        time.sleep(.5)

def update():
    global servers_found
    servers_found += 1

def thread(ipr):
    mas = masscan.PortScanner()
    mas.scan(ipr, ports='25565', arguments='--max-rate 105000')
    scan_result = json.loads(mas.scan_result)
    for ip in scan_result['scan']:
        server = JavaServer(ip, 25565)
        status = server.status()
        j = {
            'ip': server.address.host,
            'description': status.description,
            'version': status.version.name,
            'online_players': status.players.online,
            'max_players': status.players.max,
            'time_scanned': time.time()
        }
        if not rescan:
            update()
        if mycol.count_documents({'ip': server.address.host}, limit=1) != 0:
            if rescan:
                update()
            mycol.update_one({'ip': server.address.host}, j)
        else:
            mycol.insert_one(j)

if __name__ == '__main__':
    print(f"=====STARTING {'RESCAN' if rescan else 'SCAN'}=====")
    if not rescan:
        print(f'SEED: {seed}')

    rng.shuffle(ip_ranges)

    executor = concurrent.futures.ThreadPoolExecutor(threads)
    for ip_range in ip_ranges:
        idk = executor.submit(thread, ip_range)

    if not text:
        text = True
        executor2 = concurrent.futures.ThreadPoolExecutor(1)
        lol = executor2.submit(doPrintLoop)
